### PR TITLE
feat(zkstack_cli): Autocompletion support

### PR DESCRIPTION
**Pull request by manuelmauro**
## What ❔

Add autocomplete feature to zkstack:

```bash
❯ zkstack
chain              dev     -- Chain related commands                                                                                                                   
consensus                  -- Update ZKsync                                                                                                                            
containers                 -- Run containers for local development                                                                                                     
contract-verifier          -- Run contract verifier                                                                                                                    
ecosystem                  -- Ecosystem related commands                                                                                                               
explorer                   -- Run block-explorer                                                                                                                       
external-node              -- External Node related commands                                                                                                           
help                       -- Print this message or the help of the given subcommand(s)                                                                                
markdown           update  --                                                                                                                                          
portal                     -- Run dapp-portal                                                                                                                          
prover                     -- Prover related commands                                                                                                                  
server                     -- Run server
```

```bash
❯ zkstack ecosystem                   
build-transactions    -- Create transactions to build ecosystem contracts
change-default-chain  -- Change the default chain
create                -- Create a new ecosystem and chain, setting necessary configurations for later initialization
help                  -- Print this message or the help of the given subcommand(s)
init                  -- Initialize ecosystem and chain, deploying necessary contracts and performing on-chain operations
setup-observability   -- Setup observability for the ecosystem, downloading Grafana dashboards from the era-observability repo
```

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.

### CODE REVIEW

File: zkstack_cli/crates/zkstack/build.rs [hunk 1]
```diff
@@ -1,13 +1,59 @@
...
+            completion_dir.join(&completion_file),
+        )?;
+
+        // Source the completion file inside .{shell}rc
+        let shell_rc = home_dir()
+            .context("missing home directory")?
+            .join(format!(".{}rc", shell));
+
+        if shell_rc.exists() {
+            let shell_rc_content = std::fs::read_to_string(&shell_rc)?;
```
[the remark this answers is not in the document]
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs:55]: Addressed in https://github.com/matter-labs/zksync-era/pull/3097/commits/bc7d89e76cb492d29198838daf9997519319455f

File: zkstack_cli/crates/zkstack/build.rs [hunk 2]
```diff
@@ -1,13 +1,18 @@
 use std::path::PathBuf;
 
+use dirs::{config_local_dir, home_dir};
 use ethers::contract::Abigen;
+use eyre::ContextCompat;
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/build.rs:5]: Could you try to use anyhow? 
seems it's the only place where we use eyre

Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs]: Addressed in https://github.com/matter-labs/zksync-era/pull/3097/commits/ce756763c84221a3433800f56ef3c63ab5f96ffa

File: zkstack_cli/README.md [hunk 3]
```diff
@@ -30,6 +30,16 @@ zkstackup  --local
...
+### Add Autocompletion
+
+TODO: finalize section
+
+```bash
+zkstack --generate=zsh > /usr/local/share/zsh/site-functions/_zkstack
+compinit
+zkstack --<TAB>
+```
+
```
Comment (Deniallugo) [inline zkstack_cli/README.md:42]: Seems not needed anymore

File: zkstack_cli/crates/zkstack/build.rs [hunk 4]
```diff
@@ -19,3 +29,50 @@ fn main() -> eyre::Result<()> {
...
+
+                std::fs::write(
+                    shell_rc,
+                    format!(
+                        "{}\n# zkstack completion\nsource \"{}\"\n",
+                        shell_rc_content, completion_path
+                    ),
+                )?;
+            }
+        }
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/build.rs:74]: So, you never update it, once it set, right?

Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs:74]: Yes, it's enough to update the autocompletion files in the relative folders 👍

File: zkstack_cli/README.md (hunk 3 above)
Comment (manuelmauro) [reply zkstack_cli/README.md]: Addressed in https://github.com/matter-labs/zksync-era/pull/3097/commits/039e5dd7c36b4f56cd02f6f0509217db4fbc73eb

File: zkstack_cli/crates/zkstack/build.rs [hunk 5]
```diff
@@ -19,3 +29,50 @@ fn main() -> eyre::Result<()> {
...
+
+fn copy_completion_scripts() -> anyhow::Result<()> {
+    let crate_name = env!("CARGO_PKG_NAME");
+
+    // Create local config directory
+    let local_config_dir = config_local_dir().unwrap().join(crate_name);
+    std::fs::create_dir_all(&local_config_dir)?;
+
+    // Array of supported shells
+    let shells = ["bash", "zsh"];
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/build.rs:41]: do we have a special case for fish? (usually it's the most problematic one)

File: zkstack_cli/crates/zkstack/build.rs [hunk 6]
```diff
@@ -1,12 +1,22 @@
...
+    Abigen::new("ConsensusRegistry", "abi/ConsensusRegistry.json")
+        .map_err(|_| anyhow!("Failed ABI deserialization"))?
+        .generate()
+        .map_err(|_| anyhow!("Failed ABI generation"))?
+        .write_to_file(outdir.join("consensus_registry_abi.rs"))
+        .context("Failed to write ABI to file")?;
+
+    // Copy completion scripts
+    if copy_completion_scripts().is_err() {
+        println!("WARNING: It was not possible to install autocomplete scripts. Please generate them manually with `zkstack autocomplete`")
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/build.rs:18]: could we print an error as well?

File: zkstack_cli/crates/zkstack/src/commands/autocomplete.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    generate_completions(args.generator, &mut writer)?;
+
+    logger::outro(MSG_OUTRO_AUTOCOMPLETE_GENERATION);
+
+    Ok(())
+}
+
+pub fn generate_completions<G: Generator>(gen: G, buf: &mut dyn Write) -> anyhow::Result<()> {
+    let mut cmd = Inception::command();
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/src/commands/autocomplete.rs:42]: nit: you get Inception twice.

File: zkstack_cli/crates/zkstack/src/commands/dev/commands/lint.rs [hunk 8]
```diff
@@ -133,3 +149,44 @@ fn lint_contracts(shell: &Shell, ecosystem: &EcosystemConfig, check: bool) -> an
...
+        return Ok(());
+    }
+
+    // Array of supported shells
+    let shells = ["bash", "zsh"];
+
+    for shell in &shells {
+        let tmp_path = completion_folder.join(format!("_tmp_zkstack_{}", shell));
+        let tmp_file = File::create(tmp_path.clone()).context("Failed to create file")?;
+        let mut writer = BufWriter::new(tmp_file);
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/src/commands/dev/commands/lint.rs:166]: why do you need temp file? 
You can simply write to Vec<u8> and compare it later

File: zkstack_cli/crates/zkstack/src/commands/autocomplete.rs (hunk 7 above)
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/src/commands/autocomplete.rs:40]: Addresse in https://github.com/matter-labs/zksync-era/pull/3097/commits/919db1f693d49d807595afd568cdbc10b7a76e31

File: zkstack_cli/crates/zkstack/src/commands/dev/commands/lint.rs (hunk 8 above)
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/src/commands/dev/commands/lint.rs]: Nice! https://github.com/matter-labs/zksync-era/pull/3097/commits/fa44791ca3a34b6cb29471eaa1cb0d3b89e88d5d

File: zkstack_cli/crates/zkstack/src/commands/dev/commands/lint.rs [hunk 9]
```diff
@@ -133,3 +149,39 @@ fn lint_contracts(shell: &Shell, ecosystem: &EcosystemConfig, check: bool) -> an
...
+    // Array of supported shells
+    let shells = ["bash", "zsh"];
+
+    for shell in &shells {
+        let mut writer = Vec::new();
+
+        generate_completions(clap_complete::Shell::from_str(shell).unwrap(), &mut writer)
+            .context("Failed to generate autocompletion file")?;
+
+        let new = writer.iter().map(|b| *b as char).collect::<String>();
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/src/commands/dev/commands/lint.rs:169]:
```suggestion
        let new = String::from_utf8(writer)?;
```

File: zkstack_cli/crates/zkstack/build.rs (hunk 6 above)
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs]: Addressed in https://github.com/matter-labs/zksync-era/pull/3097/commits/ac22d84fe583dafad7674edec9da784a8fee6364

File: zkstack_cli/crates/zkstack/build.rs (hunk 5 above)
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs]: Added support for fish in https://github.com/matter-labs/zksync-era/pull/3097/commits/61ac014f7747429edefc275108337b0535fd1df5

File: zkstack_cli/crates/zkstack/src/commands/autocomplete.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    generate(gen, &mut cmd, cmd_name, buf);
+
+    Ok(())
+}
+
+pub trait ShellAutocomplete {
+    fn autocomplete_file_name(&self) -> &'static str;
+}
+
+impl ShellAutocomplete for clap_complete::Shell {
```
Comment (Deniallugo) [inline zkstack_cli/crates/zkstack/src/commands/autocomplete.rs:48]: Looks like unnecessary complexity

**Review by Deniallugo — COMMENTED** [review]
just nits

**Review by Deniallugo — APPROVED** [review]

File: zkstack_cli/crates/zkstack/build.rs [hunk 11]
```diff
@@ -19,3 +30,102 @@ fn main() -> eyre::Result<()> {
...
+            .context("failed to run extra configuration requirements")?;
+    }
+
+    Ok(())
+}
+
+pub trait ShellAutocomplete {
+    fn autocomplete_folder(&self) -> anyhow::Result<PathBuf>;
+    fn autocomplete_file_name(&self) -> anyhow::Result<String>;
+    fn autocomplete_extra(&self) -> anyhow::Result<()>;
```
[the remark this answers is not in the document]
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs:64]: They are extra operations needed to configure autocompletion. In this case adding a line to the `.*rc` files for bash and zsh.

File: zkstack_cli/crates/zkstack/src/commands/autocomplete.rs (hunk 10 above)
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/src/commands/autocomplete.rs]: Addressed in https://github.com/matter-labs/zksync-era/pull/3097/commits/0fee9d53e736e2147a4b4c991d7b59bbd64c6dba

File: zkstack_cli/crates/zkstack/build.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (manuelmauro) [reply zkstack_cli/crates/zkstack/build.rs]: Addressed in https://github.com/matter-labs/zksync-era/pull/3097/commits/ed0ec8452e589e1bc7ebec5349a9b5a84c9c76ee

**Review by Deniallugo — APPROVED** [review]

**Review by onyxet — APPROVED** [review]

**Pull request merged**